충남대 김규현 Step1 유효성 검사, 예외 처리 구현

HappyGogildong opened the pull request:
"카카오"문구가 포함된 상품은 담당MD와 협의 후 사용이라고 되어있는데, 담당MD와 협의하고, 상품을 승인하는 절차까지 다 구현을 해야하는 건가요? 방법을 생각해봤는데, 회원 기능 추가 시, 담당MD 역할의 회원이 관리하는 기능을 추가하면 될 것 같아 2단계에서 구현하려고
일단 예외 처리로 사용하지 못하게 해놓았습니다.

다시 브랜치 바꿨습니다

HappyGogildong committed:
<chore>(project): 상품 관리 코드 가져오기

HappyGogildong committed:
<feat, style, refactor>(exception, record, repository): 상품 유무 예외처리 추가, record 클래스 가독성 좋게 변경, containsKey로직 수정

ExceptionResponseDto : 예외 응답 반환 클래스
ProductNotFoundException : 상품 존재 x 예외
GlobalExceptionHandler : Controller의 예외 처리를 수행하는 클래스
containsKey : 쿼리문 앞에 !연산자를 제거해서 해당 id를 가진 상품이 있을 때 true를 반환하게 함

HappyGogildong committed:
<feat>(controller, requestDto): 상품 이름 유효 판단 로직 추가

공백 포함 15자 이하
특수문자 허용 : (),[],+,-,&,/,_

HappyGogildong committed:
<feat>(exception): 카카오 문구 포함 상품 예외 처리

상품명에 카카오가 포함된 상품은, 담당MD와의 사전 협의가 필요하므로 사용하지 못하도록 막음

HappyGogildong committed:
<docs>(project): 변경, 적용할 내용
README.md 에 업데이트

HappyGogildong committed:
Merge branch 'happygogildong' into step1

Rok93 commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -17,38 +19,49 @@ public ProductService(ProductRepository productRespository) {
 
     //상품 추가
     public ProductResponseDto createProduct(ProductRequestDto productRequestDto) {
+        if(productRequestDto.name().contains("카카오")){
```
> "카카오"문구가 포함된 상품은 담당MD와 협의 후 사용이라고 되어있는데, 담당MD와 협의하고, 상품을 승인하는 절차까지 다 구현을 해야하는 건가요? 방법을 생각해봤는데, 회원 기능 추가 시, 담당MD 역할의 회원이 관리하는 기능을 추가하면 될 것 같아 2단계에서 구현하려고
일단 예외 처리로 사용하지 못하게 해놓았습니다.

미션 소개 페이지에서 `기능 요구 사항에 기재되지 않은 내용은 스스로 판단하여 구현한다.` 라는 내용이 있는데요. 
이부분은 세부적인 정책이고 저 또한 알지못합니다. 😃 
규현님이 생각하시는 서비스의 정책이 있다면 거기에 맞게 구현을 하는 것이 맞다고 생각해요.  
그렇기 때문에 이 정책에 대해서는 규현님이 생각하시는 방향으로 일관성있게 에러 응답을 제공해주면 된다고 생각해요. 

다만 담당 MD에게 상품 승인을 요청하고 MD가 승인을하는 요구 절차도 기능구현을 해야하는지 질문을 주셨는데, 
이 부분은 헷갈리신다면 [API 명세](https://edu.nextstep.camp/s/UH6H8Iba/ls/TxIElkRG)를 확인해보시면 어느정도 감이 잡히실 것 같아요. 😉  

충분한 답이 됐을지 모르겠네요. 😁

Rok93 replied:
이 메서드가 호출될 때마다 `"카카오"` 라는 문자열이 생성될 것 같은데요. 
상수로 뽑아두고 재사용하도록 해보면 어떨까요?  😁

Rok93 commented on src/main/java/gift/dto/request/ProductRequestDto.java:
```diff
@@ -1,5 +1,18 @@
+    @Pattern(regexp = "^[가-힣a-zA-Z0-9 ()\\[\\]+\\-&/_]{1,15}$")
+    String name,
+    int price,
+    String imageURL) {
```
price나 imageURL에 대한 유효성 검증도 필요하지 않을까요? 🤔

Rok93 commented on src/main/java/gift/dto/request/ProductRequestDto.java:
```diff
@@ -1,5 +1,18 @@
+
+    @NotBlank
+    @Size(max = 15, message = "15자 이내로 입력해주세요")
+    @Pattern(regexp = "^[가-힣a-zA-Z0-9 ()\\[\\]+\\-&/_]{1,15}$")
```
`상품의 이름 정책`은 **도메인 정책**이라고 볼 수 있지 않을까요?  
상품을 신규 생성하고 수정하는 부분에서 동일하게 상품 이름에 대한 유효성을 검증하고있는데요. 
또 상품과 관련된 기능들이 추가되면 이 상품명 유효성 검증 코드들은 여기저기에 반복적으로 존재하게될 것 같아요. 😇

이런 정책을 도메인 객체 내부에 녹여보면 어떨까요? 😃

예를들면 지금의 경우에는 Product 객체를 생성하는 시점에 이러한 `상품 이름 정책에 맞는지 유효성을 검증`하는 로직을 추가해볼 수 있지 않을까요?

Rok93 commented on src/main/java/gift/entity/Product.java:
```diff
@@ -2,10 +2,13 @@
 
 
 import gift.dto.response.ProductResponseDto;
+import jakarta.validation.constraints.Max;
```
사용하지 않는 Import 문은 제거해주세요. 🙏

IntelliJ에서는 아래의 단축키로 자동으로 사용하지 않는 import문을 제거할 수 있어요. 
* Mac 기준: `⌃ + ⌥ + O`
* Win 기준: `Ctrl + Alt + O` 

혹은 매번 이렇게 정리해주기가 번거롭다고 느껴지신다면, IDE 설정 중에서 자동으로 사용하지 않는 import 문을 제거해주는 설정을 걸어줄 수도 있을 것 같아요. 😃

Rok93 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -29,35 +31,33 @@ public ProductController(ProductService productService) {
             HttpStatus.OK);
     }
 
     //상품 추가
```
다만 실제 프로덕트라고 생각했을 때, 이런 주석들은 조금 불필요하다고 생각해요. 👀
물론 현업에서도 아예 주석을 사용하지 않는 것은 아니에요. 
다만 가급적 주석을 먼저 쓰기보다는 코드를 통해서 최대한 표현하되 코드로는 표현하기 힘든 어려운 것들을 주석으로 남기는 편이에요. 

지금 남겨주신 주석들은 꼭 필요한거라고 생각하시나요? 😃

Rok93 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -29,35 +31,33 @@ public ProductController(ProductService productService) {
     @GetMapping("/products/{productId}")
     public ResponseEntity<ProductResponseDto> getProduct(@PathVariable long productId) {
-        return new ResponseEntity<>(productService.getProduct(productId).toResponseDto(),
+        return new ResponseEntity<>(
```
이번 미션에서 진행하신 내용은 아니지만 chaned에서 보여서 첨언을 남겨봐요. 😃

ResponseEntity에는 편의를 위한 정적 팩토리 메서드들이 구현되어있는데요. 

지금 같은 경우에는 ResponseEntity.`ok` 메서드를 대신 활용해볼 수 있을 것 같아요. 😃
```suggestion
        return ResponseEntity.ok(productService.productToResponseDto(productService.getProduct(productId)));
```
이 외에도 어떤 메서드들이 있는지 확인해보고 괜찮다고 생각되신다면 이 정적 팩토리 메서드를 활용하도록 응답을 변경해줘도 좋을 것 같아요. 😉 

<img width="791" alt="image" src="https://github.com/user-attachments/assets/dcb7480e-3dc9-45fb-8a55-2ce3aecd72ad" />

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,59 @@
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception);
+    }
+
+    @ExceptionHandler(KakaoApproveException.class)
```
만약 `ProductNotFoundException `, `MethodArgumentNotValidException`, `KakaoApproveException` 이외의 다른  예외가 발생하게되는 경우에는 어떻게 될까요? 🤔

가능하면 예기치못한 에러가 발생했을 때에는 서버 에러 응답인 500 에러 응답을 내려주도록해주면 조금 더 깔끔한 일관된 응답을 해줄 수 있지 않을까요? 😁

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,59 @@
+                .append(fieldError.getDefaultMessage());
+        }
+
+        ExceptionResponseDto exception = new ExceptionResponseDto(
```
사소하지만 ExceptionResponse 보다는 `ErrorResponse` 와 같은 네이밍은 어떨까요?

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,59 @@
+
+    @ResponseBody
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ExceptionResponseDto> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+        StringBuilder sb = new StringBuilder();
+
+        for (FieldError fieldError : e.getBindingResult().getFieldErrors()) {
+            sb.append(fieldError.getField())
+                .append(":")
+                .append(fieldError.getDefaultMessage());
+        }
```
이렇게 에러 메시지를 보여준다면    예를들어 2가지 이상의 유효성 검증에 걸리게되면 어떻게 표기가 되는걸까요? 👀
두 예외 메시지가 연결돼서 보이지 않을까요?  

제가 생각한게 맞다면 가독성이 조금 떨어질 것 같다고 생각되는 것 같아요. 🤔

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,59 @@
+    @ResponseBody
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ExceptionResponseDto> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+        StringBuilder sb = new StringBuilder();
+
+        for (FieldError fieldError : e.getBindingResult().getFieldErrors()) {
+            sb.append(fieldError.getField())
+                .append(":")
+                .append(fieldError.getDefaultMessage());
+        }
+
+        ExceptionResponseDto exception = new ExceptionResponseDto(
+            sb.toString(),
+            400,
+            LocalDateTime.now()
+        );
+
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception);
```
아래와 같이 변경해보는건 어떨까요? 🤗 

```suggestion
        String errorMessage = e.getBindingResult().getFieldErrors()
            .stream()
            .map(fieldError -> fieldError.getField() + ":" + fieldError.getDefaultMessage())
            .collect(Collectors.joining(", "));

        ExceptionResponseDto exception = new ExceptionResponseDto(
            errorMessage,
            HttpStatus.BAD_REQUEST.value(),
            LocalDateTime.now()
        );
```

Rok93 commented on src/main/java/gift/repository/ProductJdbcRepository.java:
```diff
@@ -71,7 +71,7 @@ public void delete(long productId) {
 
     @Override
     public boolean containsKey(long id) {
```
네이밍이 조금 이상한 것 같은데요. 🤔
이건 `containsKey`를 의도하신게 맞을까요? 

product의 식별자 id로 조회하는 것이니 key가 포함됐는지 아닌지 여부를 확인하는 메서드는 아닌 것 같아서요. 👀

존재여부(= `exist`)에 가깝지 않을까요?

Rok93 commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -17,38 +19,49 @@ public ProductService(ProductRepository productRespository) {
+
     //상품 단 건 조회
     public Product getProduct(long productId) {
+        if(!containsProduct(productId)){throw new ProductNotFoundException("상품을 찾을 수 없습니다");}
```
IDE의 자동 정렬을 습관적으로 활용해주세요. 

* Mac 기준: `⌥ + ⌘ + L`
* Win 기준: `Ctrl + Alt + L`

마찬가지로 이것도 IDE에서 자동으로 동작하도록 셋팅할 수 있으니 같이 확인해서 설정해보셔도 좋을 것 같습니다.

Rok93 reviewed (changes requested):
규현님 1단계 구현 잘해주셨습니다. 👍👍 
몇몇 코멘트 남겨두었는데 확인 부탁드릴게요. 🤗
미션 진행하시면서 어려운 부분있으면 언제든지 DM 주세요. 😉

HappyGogildong committed:
<refactor>(project): 피드백 내용 반영

1. 저는 관리자 페이지를 담당MD가 사용하고
RESTapi는 일반 사용자를 위한 것으로 구분하였습니다.
승인 방식은, 담당MD가 물품을 사전등록하는 것으로 하고,
담당MD의 역할을 가진 사용자만 카카오 문구를 넣을 수 있게 만듭니다
회원 기능 구현 시 두 유형의 사용자를 구분하는 로직을 추가하겠습니다

2. price, URL 유효성 검증 로직을 추가했습니다.
상품 최대 가격은, 상품명에 대해 피드백 주셨던 것처럼
도메인 정책에 포함되는 것 같아 Product 클래스에서 검증하게 했습니다

3. 상품명 정책을 Product 객체 내부에서 검사하게 했습니다. 이 때 던지는 예외가 기존 어노테이션이 던지던 예외와 달라 GlobalExceptionHandler에 메서드를 추가했습니다. 기존에 던지던 예외는
MethodArgumentNotValidException인데, 이 예외는
@Valid등의 Dto 유효성 검증에 실패할 때 Controller에서 발생시키는 예외로 역할이 구분되어 있어, Product 도메인에서 도메인 정책을 위해 발생시키는 예외는 다른 예외를 호출하게 했습니다.

4. 사용하지 않는 import문을 제거하고
자동 정렬 기능을 켰습니다.

5. 불필요한 주석을 제거했습니다.
Controller의 기능은 메서드, URL을 보고
API명세서와 대조해 충분히 확인 가능하다 생각해 지웠고,
Service와 repository는 메서드명을 보고 유추할 수 있다고 생각했습니다.

6. handleMethodArgumentNotValidException
의 구조를 추천해주신 stream기반으로 변경했습니다

7. 가독성을 위해 ExceptionResponseDto의 errorMessage필드를 String에서 List<String>으로 바꾸어 에러 요소마다 구분되게 하였습니다.

8. 테스트 코드를 추가했습니다

9. "카카오" 문자열은 Java의 String literal이고,
String Constant Pool에 저장되어 이 리터럴을 여러 번 사용할 때 동일한 객체가 재사용된다는 내용을 찾아보았습니다. 저는 피드백 주신 내용을
contains 메서드가 호출될 때마다 "카카오"문자열 객체가 생성된다는 뜻으로 받아들였는데요, 제가 찾아본 내용과 다른 것 같아 다시 질문드립니다.
https://deveric.tistory.com/123
https://www.baeldung.com/java-string-pool
https://hyunsb.tistory.com/66
https://codepumpkin.com/string-pool-java/

HappyGogildong committed:
Merge remote-tracking branch 'origin/step1' into step1

Rok93 commented on README.md:
```diff
@@ -1 +1,11 @@
\ No newline at end of file
+# spring-gift-wishlist
+
+### 1단계 유효성, 예외 처리
```
1단계 진행이 늦어서 2단계 리뷰요청 마감일에 맞게 리뷰요청을 못주셔겠군요. 🥲
매니저님께 리뷰어와 논의가 되었으니 2단계 리뷰 요청을 내일까지 진행하겠다고 공유해주시고 2단계 진행 내일까지 진행 부탁드려요.

Rok93 reviewed (approved):
규현님 피드백 반영 잘해주셨습니다. 
2단계 시작이 조금 늦었는데, 지금부터라도 빠르게 달려보시죠. 💪💪
